src/audio/codec_adapter/codec/generic.c

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Marcin Rajwa <marcin.rajwa@linux.intel.com>
+
+/*
+ * \file generic.c
+ * \brief Generic Codec API
+ * \author Marcin Rajwa <marcin.rajwa@linux.intel.com>
+ *
+ */
+
+#include <sof/audio/codec_adapter/codec/generic.h>
+
+/*****************************************************************************/
+/* Local helper functions						     */
+/*****************************************************************************/
+static int validate_config(struct codec_config *cfg);
+
+int
+codec_load_config(struct comp_dev *dev, void *cfg, size_t size,
+		  enum codec_cfg_type type)
+{
+	int ret;
+	struct codec_config *dst;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct codec_data *codec = &cd->codec;
+
+	comp_dbg(dev, "codec_load_config() start");
+
+	if (!dev || !cfg || !size) {
+		comp_err(dev, "codec_load_config(): wrong input params! dev %x, cfg %x size %d",
+			 (uint32_t)dev, (uint32_t)cfg, size);
+		return -EINVAL;
+	}
+
+	dst = (type == CODEC_CFG_SETUP) ? &codec->s_cfg :
+					  &codec->r_cfg;
+
+	if (!dst->data) {
+		dst->data = rballoc(0, SOF_MEM_CAPS_RAM, size);
+	} else if (dst->size != size) {
+		rfree(dst->data);
+		dst->data = rballoc(0, SOF_MEM_CAPS_RAM, size);
+	}
+	if (!dst->data) {
+		comp_err(dev, "codec_load_config(): failed to allocate space for setup config.");
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	ret = memcpy_s(dst->data, size, cfg, size);
+	assert(!ret);
+	ret = validate_config(dst->data);
+	if (ret) {
+		comp_err(dev, "codec_load_config(): validation of config failed!");
+		ret = -EINVAL;
+		goto err;
+	}
+
+	/* Config loaded, mark it as valid */
+	dst->size = size;
```

Comment:
Might want to move this size assignment next to the allocation, otherwise we might get memory management surprises later on.

Reply:
I don't get this proposal. You start with `size = 0` and if allocation + config load + validation all went well you set `size` as a mark of valid config. Otherwise you just free the allocated memory and `size` remains 0.